### PR TITLE
Fix main provider flow to try other bearer token providers if exchang…

### PR DESCRIPTION
**Pull request by zarenner**
…ing for a session token fails.

Add exception handling to continue trying other bearer token providers if one throws.

### CODE REVIEW

**Comment by zarenner** [comment]
#41

File: CredentialProvider.Microsoft/Resources.resx [hunk 1]
```diff
@@ -381,10 +365,25 @@ Cache Location
   <data name="BuildTaskEndpointEnvVarError" xml:space="preserve">
     <value>This credential provider must be run under the Team Build tasks for NuGet with external endpoint credentials</value>
   </data>
-  <data name="AdalAcquireTokenWIAFailed" xml:space="preserve">
-    <value>Failed to acquire token using Windows Integrated Authentication</value>
+  <data name="AcquireBearerTokenFailed" xml:space="preserve">
+    <value>Failed to acquire token using '{0}'</value>
+  </data>
```
Comment (salvmd) [inline CredentialProvider.Microsoft/Resources.resx]: not used?

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs [hunk 2]
```diff
@@ -55,82 +51,67 @@ public override async Task<bool> CanProvideCredentialsAsync(Uri uri)
 
         public override async Task<GetAuthenticationCredentialsResponse> HandleRequestAsync(GetAuthenticationCredentialsRequest request, CancellationToken cancellationToken)
         {
-            // If this is a retry, let's try without sending a retry request to the underlying bearerTokenProvider
```
Comment (salvmd) [inline CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs]: It doesn't look like we're doing this anymore (the line below was also removed: "var responseWithNoRetry = await HandleRequestAsync(GetNonRetryRequest(request), cancellationToken);").

I'm not sure I understand why we were doing this recursive call with IsRetry: false before. Was this change on purpose?

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/BearerTokenProvider.cs [hunk 3]
```diff
@@ -13,129 +12,133 @@
 
 namespace NuGetCredentialProvider.CredentialProviders.Vsts
 {
-    public class BearerTokenProvider : IBearerTokenProvider
+    /// <summary>
+    /// Acquire an AAD token via the ADAL cache
+    /// </summary>
+    public class ADALCacheBearerTokenProvider : IBearerTokenProvider
```
[the remark this answers is not in the document]
Comment (zarenner) [reply CredentialProvider.Microsoft/CredentialProviders/Vsts/BearerTokenProvider.cs]: While I personally disagree with that [style guide point](https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-1.1/141e06ef(v=vs.71)), you're right. Will change.

(sidenote: The new netstandard [guidance](https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/general-naming-conventions#using-abbreviations-and-acronyms) *really* doesn't like acronyms)

File: CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs [hunk 4]
```diff
@@ -16,25 +16,21 @@ namespace NuGetCredentialProvider.CredentialProviders.Vsts
...
-
-        public VstsCredentialProvider(ILogger logger, IAuthUtil authUtil, IBearerTokenProvider bearerTokenProvider)
+        private readonly IBearerTokenProvidersFactory bearerTokenProvidersFactory;
+        private readonly IAzureDevOpsSessionTokenFromBearerTokenProvider vstsSessionTokenProvider;
+
+        public VstsCredentialProvider(
+            ILogger logger,
+            IAuthUtil authUtil,
+            IBearerTokenProvidersFactory bearerTokenProvidersFactory,
+            IAzureDevOpsSessionTokenFromBearerTokenProvider vstsSessionTokenProvider)
```
[the remark this answers is not in the document]
Comment (zarenner) [reply CredentialProvider.Microsoft/CredentialProviders/Vsts/VstsCredentialProvider.cs]: Gonna leave this for now - was about to merge and there's still a few other VSTS's hanging around. Let's do a VSTS cleanup pass though as part of the next round of work here.
